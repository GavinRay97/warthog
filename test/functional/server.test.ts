import * as fs from 'fs';
import * as path from 'path';

import { get, GetResponse } from '../../src/core/http';
import { Server } from '../../src/core/server';
import { logger } from '../../src/core/logger';

// import { Binding } from '../../test/generated/binding';
import { getTestServer } from '../test-server';
import { KitchenSink } from '../modules';

let server: Server<any>;
// let binding: Binding;
let binding: any;

describe('server', () => {
  beforeEach(() => {
    jest.setTimeout(20000);
  });

  // Make sure to clean up server
  beforeAll(async done => {
    jest.setTimeout(20000);
    await cleanUpTestData();

    try {
      server = getTestServer({
        apolloConfig: { playground: false }
      });

      await server.start();

      binding = (await server.getBinding()) as unknown;
    } catch (error) {
      logger.error(error);
      process.exit(1);
    }

    const kitchenSink = await createKitchenSink(binding);
    await createDishes(binding, kitchenSink.id);

    done();
  });

  // Make sure to clean up server
  afterAll(async done => {
    await cleanUpTestData();
    await server.stop();
    done();
  });

  it('disables playground properly using apollo config options', async () => {
    const response: GetResponse = await get(server.getGraphQLServerUrl());

    expect(response.statusCode).toEqual(400);
    expect(response.body).toContain('GET query missing');
  });

  it('queries effectively', async () => {
    expect(
      await binding.query.kitchenSinks(
        { skip: 0, orderBy: 'createdAt_DESC', limit: 5 },
        `{
          stringField
          emailField
          integerField
          booleanField
          floatField
          dishes {
            name
          }
          createdById
          updatedById
          version
        }`
      )
    ).toMatchSnapshot();
  });
});

async function createKitchenSink(binding: any): Promise<KitchenSink> {
  return binding.mutation.createKitchenSink(
    {
      data: {
        stringField: 'My String',
        emailField: 'hi@warthog.com',
        integerField: 123,
        booleanField: true,
        floatField: 123.456
      }
    },
    `{ id }`
  );
}

async function createDishes(binding: any, kitchenSinkId: string): Promise<KitchenSink> {
  const data = Array.from({ length: 20 }, (v, i) => i).map(item => {
    return {
      name: `Dish ${item}`,
      kitchenSinkId
    };
  });
  let dishes;

  try {
    dishes = await binding.mutation.createManyDishs({ data }, `{ id name createdById }`);
  } catch (error) {
    logger.error(error);
  }

  return dishes;
}

async function cleanUpTestData() {
  try {
    fs.unlinkSync(path.join(__dirname, '../../warthog.sqlite.tmp'));
  } catch (error) {
    // console.error('Error cleaning up test data', error);
  }

  return;
}
