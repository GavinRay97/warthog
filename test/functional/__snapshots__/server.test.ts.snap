// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server queries deeply nested objects 1`] = `
Array [
  Object {
    "booleanField": true,
    "createdById": "abc123",
    "dishes": Array [
      Object {
        "name": "Dish 0",
      },
      Object {
        "name": "Dish 1",
      },
      Object {
        "name": "Dish 2",
      },
      Object {
        "name": "Dish 3",
      },
      Object {
        "name": "Dish 4",
      },
      Object {
        "name": "Dish 5",
      },
      Object {
        "name": "Dish 6",
      },
      Object {
        "name": "Dish 7",
      },
      Object {
        "name": "Dish 8",
      },
      Object {
        "name": "Dish 9",
      },
      Object {
        "name": "Dish 10",
      },
      Object {
        "name": "Dish 11",
      },
      Object {
        "name": "Dish 12",
      },
      Object {
        "name": "Dish 13",
      },
      Object {
        "name": "Dish 14",
      },
      Object {
        "name": "Dish 15",
      },
      Object {
        "name": "Dish 16",
      },
      Object {
        "name": "Dish 17",
      },
      Object {
        "name": "Dish 18",
      },
      Object {
        "name": "Dish 19",
      },
    ],
    "emailField": "hi@warthog.com",
    "floatField": 123.456,
    "integerField": 123,
    "stringField": "My String",
    "updatedById": null,
    "version": 1,
  },
]
`;

exports[`server string query: contains \`a\` (upper or lower) 1`] = `
Array [
  Object {
    "stringField": "Trantow",
  },
  Object {
    "stringField": "VOLUPTATUM NULLA",
  },
  Object {
    "stringField": "maximillia",
  },
  Object {
    "stringField": "QUAM CULPA",
  },
  Object {
    "stringField": "Madeline",
  },
  Object {
    "stringField": "NAKIA",
  },
  Object {
    "stringField": "nakia",
  },
  Object {
    "stringField": "Nakia",
  },
  Object {
    "stringField": "sed praesentium",
  },
  Object {
    "stringField": "amber",
  },
  Object {
    "stringField": "delfinaz",
  },
  Object {
    "stringField": "SHEMAR",
  },
  Object {
    "stringField": "stark",
  },
  Object {
    "stringField": "elda",
  },
  Object {
    "stringField": "Damien",
  },
  Object {
    "stringField": "earum nesciunt",
  },
  Object {
    "stringField": "illum atque",
  },
  Object {
    "stringField": "schowalter",
  },
  Object {
    "stringField": "MAIORES ADIPISCI",
  },
  Object {
    "stringField": "VELIT ADIPISCI",
  },
  Object {
    "stringField": "AMET QUI",
  },
  Object {
    "stringField": "maiores praesentium",
  },
  Object {
    "stringField": "KAELYN",
  },
  Object {
    "stringField": "alias sint",
  },
  Object {
    "stringField": "molestiae praesentium",
  },
  Object {
    "stringField": "Okuneva",
  },
  Object {
    "stringField": "hartmann",
  },
  Object {
    "stringField": "raquel",
  },
  Object {
    "stringField": "quas fugit",
  },
  Object {
    "stringField": "quia et",
  },
  Object {
    "stringField": "ut consequatur",
  },
  Object {
    "stringField": "iusto perspiciatis",
  },
  Object {
    "stringField": "HAMILL",
  },
  Object {
    "stringField": "LABORE CULPA",
  },
  Object {
    "stringField": "VERITATIS EIUS",
  },
  Object {
    "stringField": "HELGA",
  },
  Object {
    "stringField": "ipsum voluptas",
  },
  Object {
    "stringField": "linnea",
  },
  Object {
    "stringField": "madison",
  },
  Object {
    "stringField": "NUMQUAM ALIQUAM",
  },
  Object {
    "stringField": "macy",
  },
  Object {
    "stringField": "GERLACH",
  },
  Object {
    "stringField": "padberg",
  },
  Object {
    "stringField": "ENIM DICTA",
  },
  Object {
    "stringField": "amalia",
  },
  Object {
    "stringField": "QUAS PERSPICIATIS",
  },
  Object {
    "stringField": "Williamson",
  },
  Object {
    "stringField": "schamberger",
  },
  Object {
    "stringField": "Jared",
  },
  Object {
    "stringField": "QUIDEM FUGIAT",
  },
]
`;

exports[`server string query: ends with \`z\` (upper or lower) 1`] = `
Array [
  Object {
    "stringField": "LUEILWITZ",
  },
  Object {
    "stringField": "delfinaz",
  },
  Object {
    "stringField": "horowitz",
  },
]
`;

exports[`server string query: exact match (Nakia) 1`] = `
Array [
  Object {
    "stringField": "Nakia",
  },
]
`;

exports[`server string query: starts with \`b\` (upper or lower) 1`] = `
Array [
  Object {
    "stringField": "BOSCO",
  },
  Object {
    "stringField": "BERENICE",
  },
  Object {
    "stringField": "billie",
  },
  Object {
    "stringField": "Block",
  },
  Object {
    "stringField": "BAUCH",
  },
]
`;
